// wikibase-sdk/lib/helpers/simplify.js
const { labels, descriptions, aliases, lemmas, glosses } = require('wikibase-sdk/lib/helpers/simplify_text_attributes')

const {
  simplifyClaim: claim,
  simplifyPropertyClaims: propertyClaims,
  simplifyClaims: claims,
  simplifyQualifier: qualifier,
  simplifyPropertyQualifiers: propertyQualifiers,
  simplifyQualifiers: qualifiers,
  simplifyReferences: references,
} = require('wikibase-sdk/lib/helpers/simplify_claims')

const { simplifyForm: form, simplifyForms: forms } = require('wikibase-sdk/lib/helpers/simplify_forms')
const { simplifySense: sense, simplifySenses: senses } = require('wikibase-sdk/lib/helpers/simplify_senses')

const sitelinks = require('wikibase-sdk/lib/helpers/simplify_sitelinks')
const sparqlResults = require('wikibase-sdk/lib/helpers/simplify_sparql_results')

module.exports = {
  labels,
  descriptions,
  aliases,
  claim,
  propertyClaims,
  claims,
  qualifier,
  propertyQualifiers,
  qualifiers,
  references,
  sitelinks,

  // Aliases
  snak: claim,
  propertySnaks: propertyClaims,
  snaks: claims,

  // Lexemes
  lemmas,
  glosses,
  form,
  forms,
  sense,
  senses,

  sparqlResults

  // Set in ./simplify_entity
  // entity,
  // entities,
}
